hibernate-core/src/main/java/org/hibernate/mapping/Component.java

```diff
@@ -209,13 +212,28 @@ public void setDynamic(boolean dynamic) {
 
 	@Override
 	public Type getType() throws MappingException {
-		// TODO : temporary initial step towards HHH-1907
-		final ComponentMetamodel metamodel = new ComponentMetamodel(
-				this,
-				getMetadata().getMetadataBuildingOptions()
-		);
-		final TypeFactory factory = getMetadata().getTypeConfiguration().getTypeResolver().getTypeFactory();
-		return isEmbedded() ? factory.embeddedComponent( metamodel ) : factory.component( metamodel );
+		// Resolve the type of the value once and for all as this operation generates a proxy class
+		// for each invocation.
+		// Unfortunately, there's no better way of doing that as none of the classes are immutable and
+		// we can't know for sure the current state of the property or the value.
+		Type localType = type;
+
+		if ( localType == null ) {
+			synchronized ( this ) {
+				if ( type == null ) {
+					// TODO : temporary initial step towards HHH-1907
+					final ComponentMetamodel metamodel = new ComponentMetamodel(
+							this,
+							getMetadata().getMetadataBuildingOptions()
+					);
+					final TypeFactory factory = getMetadata().getTypeConfiguration().getTypeResolver().getTypeFactory();
+					localType = isEmbedded() ? factory.embeddedComponent( metamodel ) : factory.component( metamodel );
+					type = localType;
+				}
+			}
+		}
```

Comment:
This is a bit unfortunate, but I must admit I can't think of any other solution, considering:

* there's no obvious place in the bootstrap process from where we could call some `init` method on this class
* it's not obvious what data is required to initialize the type either, so we can't just replace various setters with a big `setTypeInfo(Something, Somethingelse, ...)` setter.

I guess there are ways to solve this by changing how `org.hibernate.tuple.Instantiator` is... instantiated, but that would require changing `AbstractComponentTuplizer` and all its subclasses, and maybe also `ComponentTuplizerFactory` and `ComponentMetamodel`, to rely on a global cache at some level. But that sounds a bit dangerous, and I'm not sure it's doable without changing publicly exposed methods...

So... +1.

Reply:
Yeah, my thoughts exactly.

Another line of thought is to consider this method will be called at least once during the bootstrap in a mono-threaded environment before being used in a multi-threaded one. Thus we wouldn't need the volatile/synchronized. I'm not sure I want to go that way and expose us to weird multi-threading issues if one day, this method is not called anymore during bootstrap for whatever reason.